src/builtins/fish_indent.rs

```diff
@@ -0,0 +1,1200 @@
+    PROGRAM_NAME.set(L!("fish_indent")).unwrap();
+    panic_handler(throwing_main)
+}
+
+fn throwing_main() -> i32 {
+    // TODO: Duplicated with fish_key_reader
+    use crate::io::FdOutputStream;
+    use crate::io::IoChain;
+    use crate::io::OutputStream::Fd;
+    use libc::{STDERR_FILENO, STDOUT_FILENO};
+
+    topic_monitor_init();
+    threads::init();
+
+    let mut out = Fd(FdOutputStream::new(STDOUT_FILENO));
+    let mut err = Fd(FdOutputStream::new(STDERR_FILENO));
+    let io_chain = IoChain::new();
+    let mut streams = IoStreams::new(&mut out, &mut err, &io_chain);
+    // Using the user's default locale could be a problem if it doesn't use UTF-8 encoding. That's
+    // because the fish project assumes Unicode UTF-8 encoding in all of its scripts.
+    //
+    {
+        let s = CString::new("").unwrap();
+        unsafe { libc::setlocale(LC_ALL, s.as_ptr()) };
+    }
+    env_init(None, true, false);
+    let args: Vec<WString> = std::env::args_os()
+        .map(|osstr| str2wcstring(osstr.as_bytes()))
+        .collect();
+
+    // Only set these here so you can't set them via the builtin.
+    if let Some(features_var) = EnvStack::globals().get(L!("fish_features")) {
+        for s in features_var.as_list() {
+            future_feature_flags::set_from_string(s.as_utfstr());
+        }
+    }
+
+    do_indent(&mut streams, args)
+}
+
+pub fn fish_indent(_parser: &Parser, streams: &mut IoStreams, args: &mut [&wstr]) -> Option<c_int> {
+    let args = args.iter_mut().map(|x| x.to_owned()).collect();
+    Some(do_indent(streams, args))
+}
+
+fn do_indent(streams: &mut IoStreams, args: Vec<WString>) -> i32 {
+    // Types of output we support
+    #[derive(Eq, PartialEq)]
+    enum OutputType {
+        PlainText,
+        File,
+        Ansi,
+        PygmentsCsv,
+        Check,
+        Html,
+    }
+
+    let mut output_type = OutputType::PlainText;
+    let mut output_location = L!("");
+    let mut do_indent = true;
+    let mut only_indent = false;
+    let mut only_unindent = false;
+    // File path for debug output.
+    let mut debug_output = None;
+
+    let short_opts: &wstr = L!("+d:hvwicD:");
+    let long_opts: &[WOption] = &[
+        wopt(L!("debug"), ArgType::RequiredArgument, 'd'),
+        wopt(L!("debug-output"), ArgType::RequiredArgument, 'o'),
+        wopt(L!("debug-stack-frames"), ArgType::RequiredArgument, 'D'),
+        wopt(L!("dump-parse-tree"), ArgType::NoArgument, 'P'),
+        wopt(L!("no-indent"), ArgType::NoArgument, 'i'),
+        wopt(L!("only-indent"), ArgType::NoArgument, '\x04'),
+        wopt(L!("only-unindent"), ArgType::NoArgument, '\x05'),
+        wopt(L!("help"), ArgType::NoArgument, 'h'),
+        wopt(L!("version"), ArgType::NoArgument, 'v'),
+        wopt(L!("write"), ArgType::NoArgument, 'w'),
+        wopt(L!("html"), ArgType::NoArgument, '\x01'),
+        wopt(L!("ansi"), ArgType::NoArgument, '\x02'),
+        wopt(L!("pygments"), ArgType::NoArgument, '\x03'),
```

Comment:
This is used by our docs - we highlight the code samples by shelling out to `fish_indent`.

That would have to be adjusted to call `fish -c 'fish_indent $argv' args` if we decide to drop the second binary.